flfm/app/main.py

```diff
@@ -0,0 +1,470 @@
+    if data is None:
+        return no_update
+
+    n_frames = len(data)
+    center_frame = _center_frame(n_frames)
+    default_slider_value = center_frame
+    fig = plot_image(data[center_frame, :, :])
+    children = [
+        dcc.Graph(id=dict(type="image-graph", index=id), figure=fig, responsive=True, style={"height": "100vh"}),
+        html.Label("Select Image Frame/Depth"),
+        dcc.Slider(
+            0,
+            len(data) - 1,
+            step=1,
+            value=default_slider_value,
+            id=dict(type="image-slider", index=id),
+            disabled=n_frames <= 1,
+            tooltip={
+                "always_visible": n_frames > 1,
+                "template": slider_template_str(default_slider_value, center_frame, app_settings.DEPTH_STEP_SIZE),
+            },
+        ),
+        html.Label("Color Scale"),
+        dcc.Dropdown(
+            id=dict(type="color-scale", index=id),
+            options=colorscales,
+            value=app_settings.IMSHOW_COLOR_SCALE,
+        ),
+    ]
+    return children, n_frames
+
+
+def upload_file(contents, filename):
+    """Read in data/image files."""
+    if contents is not None:
+        logger.info(f"Uploading {filename}...")
+        content_type, content_string = contents.split(",", maxsplit=1)
+        decoded = base64.b64decode(content_string)
+        bytestream = io.BytesIO(decoded)
+        data = flfm_io.open(bytestream)
+        logger.info(f"Uploading {filename} completed.")
+        return data
+
+
+@callback(
+    Output(dict(type="display", index="reconstruction"), "children", allow_duplicate=True),
+    Input("run-reconstruction", "n_clicks"),
+    Input("rl-iters", "value"),
+    State("center_x", "value"),
+    State("center_y", "value"),
+    State("radius", "value"),
+    prevent_initial_call=True,
+)
+def reconstruct(n_clicks, rl_iters, center_x, centre_y, radius):
+    """RUN FLFM reconstruction and return reconstructed image."""
+    global image_data
+    if n_clicks == 0:
+        return no_update
+
+    if image_data["psf"] is None or image_data["image"] is None:
+        raise no_update
+
+    id = "reconstruction"
+    image_data["uncropped_reconstruction"] = flfm_restoration.richardson_lucy(
+        image_data["image"], image_data["psf"], num_iter=rl_iters
+    )
+    image_data[id] = flfm.util.crop_and_apply_circle_mask(
+        image_data["uncropped_reconstruction"],
+        center=(center_x, centre_y),
+        radius=radius,
+    )
+    children, _n_frames = _display_image_with_slider(image_data[id], id=id)
+    return children
+
+
+@callback(
+    Output(dict(type="image-graph", index="psf"), "figure", allow_duplicate=True),
+    Output("normalize-psf", "disabled"),
+    Input("normalize-psf", "n_clicks"),
+    State(dict(type="image-slider", index="psf"), "value"),
+    State(dict(type="color-scale", index="psf"), "value"),
+    prevent_initial_call=True,
+)
+def normalize_psf(n_clicks, frame, color_scale):
```

Comment:
Can you add some type hints here if the types are consistent and reasonable to reference?